File: examples/blotter/backend/run-testdf.py
import ipdb
import sys, asyncio
from KVAN import fuargs, topdir
topdir.setup_syspath()
import pybx, pandas as pd, threading
import pybx_comm
import random
pybx.import_pybx("Blotter")
pybx.import_pybx("Utils")

class ObserverI(Blotter.Observer):
    async def show(self, df):
        print("ObserverI::show:", df)

cities = [
    ['Boston', 'MA'],
    ['Worcester', 'MA'],
    ['New York', 'NY'],
    ['Albany', 'NY']
]

class DFTestI(Blotter.DFTest):
    def __init__(self):
        self.subscribers = []
        self.df_lock = threading.Lock()
        self.df = pd.DataFrame.from_records(cities, columns = ['city', 'state'])
        self.df['temp'] = 25
        self.c = 0

    # see about data race example and GIL role -> https://stackoverflow.com/a/33435680/1181482
    async def update_task(self):
        print("staring update_task")
        while 1:
            with self.df_lock:
                print("DFTestI::update", self.c)
                for ii, rr in self.df.iterrows():
                    incr = (random.random() - 0.5) * 0.45
                    if abs(incr) > 0.1:
                        self.c += 1
                        self.df.loc[ii, 'temp'] = rr['temp'] + incr
            await self.publish()
            await asyncio.sleep(2)
        
    async def get_df(self):
        with self.df_lock:
            df_ret = Utils.DataFrame(columns = list(self.df.columns), dataframeJSON = self.df.to_json(orient = 'records'))
            return Blotter.DFWUPC(df = df_ret, update_c = self.c)

    async def subscribe(self, rop):
        print("DFTestI::subscribe:", rop)
        #ipdb.set_trace()
        with self.df_lock:
            res_rop = self.subscribers.append(rop)
            print("res_rop: ", res_rop)

    async def publish(self):
        with self.df_lock:
            for rop in self.subscribers:
                try:
                    #ipdb.set_trace()
                    df_o = Utils.DataFrame(columns = list(self.df.columns), dataframeJSON = self.df.to_json(orient = 'records'))
                    j = Blotter.DFWUPC(df = df_o, update_c = self.c)
                    await rop.show(j)
                except Exception as e:
                    print("rop.show failed", rop)
                    print(e)
            
async def test_coro():
    comm = pybx_comm.Communicator()
    testdf_rop = await comm.get_rop(Blotter.DFTest, "ws://localhost:8080/", "test_df")

    df = await testdf_rop.get_df()
    print("dataframe:", df)

@fuargs.action
def test():
    asyncio.get_event_loop().run_until_complete(test_coro())
    
async def test_subscriber_coro():
    comm = pybx_comm.Communicator()
    testdf_rop = await comm.get_rop(Blotter.DFTest, "ws://localhost:8080/", "test_df")

    subscriber_o = ObserverI()
    subscriber_rop = comm.add_object(subscriber_o, "uu")
    res_rop = await testdf_rop.subscribe(subscriber_rop)
    print("RES_ROP:", res_rop)

@fuargs.action
def test_subscriber():
    asyncio.get_event_loop().run_until_complete(test_subscriber_coro())
    asyncio.get_event_loop().run_forever()
    
async def run_coro():
    port = 8080
    host = "0.0.0.0"
    comm = pybx_comm.Communicator()
    comm.set_listen_port(port, host)
    test_o = DFTestI()
    comm.add_object(test_o, "test_df")

    update_task = test_o.update_task()
    print("update_task:", update_task, type(update_task))
    ws_server_task = comm.start_server()
    print("ws_server_task:", ws_server_task, type(ws_server_task))        
    if 1:
        asyncio.create_task(update_task)
        await ws_server_task
    else:
        await asyncio.gather(update_task, ws_server_task)
        #asyncio.gather(update_task, ws_server_task, return_exceptions=True)

    print("python server start")    

@fuargs.action
def run():
    asyncio.get_event_loop().run_until_complete(run_coro())
    asyncio.get_event_loop().run_forever()
    
if __name__ == "__main__":
    fuargs.exec_actions(sys.argv[1:])

File: examples/blotter/backend/run-testdf.cpp
#include <iostream>
#include <thread>
using namespace std;

#include <kvan/fuargs.h>
#include <kvan/fjson-io.h>
#include <libpybx-cpp/communicator.h>
#include "Blotter_pybx.h"
#include <unistd.h>

class ObserverI : public pybx::Blotter::Observer {
public:
  void show(pybx::Blotter::DFWUPC df) override {
    cout << "show: " << df.df.dataframeJSON << endl;
  }
};

class DFTestI : public pybx::Blotter::DFTest {
private:
  mutex obj_lock;
  vector<pybx::Blotter::Observer_rop> rops;
  
public:
  struct Rec {
    string city, state;
    double chance_of_rain{0.0};
  };
  vector<Rec> v;
  
  DFTestI() {
    this->v = vector<Rec>{
      Rec{.city = "Ashland", .state = "MA", .chance_of_rain = 0.2},
      Rec{.city = "Hopkinton", .state = "MA", .chance_of_rain = 0.23},
      Rec{.city = "Hartford", .state = "CT", .chance_of_rain = 0.22},
      Rec{.city = "Greenwich", .state = "CT", .chance_of_rain = 0.26}
    };
  }
  
  pybx::Blotter::DFWUPC get_df() override {
    lock_guard<mutex> l(obj_lock);
    ostringstream out;
    auto columns = to_fjson(out, v);
    pybx::Blotter::DataFrame df{.columns = columns, .dataframeJSON = out.str()};
    return pybx::Blotter::DFWUPC{.df = df, .update_c = -1};
  }

  pybx::Blotter::Observer_rop subscribe(pybx::Blotter::Observer_rop rop) override {
    cout << "DFTestI::subscribe" << endl;
    lock_guard<mutex> l(obj_lock);
    rops.push_back(rop);
    return rop;
  }

  void update_thread()
  {
    cout << "starting update thread" << endl;
    int i = 0;
    while (1) {
      cout << "update thread " << i++ << endl;
      {
	lock_guard<mutex> l(obj_lock);
	for (auto& el: v) {
	  el.chance_of_rain += 0.001;
	}
      }
      this->publish();
      sleep(1);
    }
  }
  
  void publish()
  {
    lock_guard<mutex> l(obj_lock);	
    ostringstream out;
    auto columns = to_fjson(out, v);
    pybx::Blotter::DataFrame df{.columns = columns, .dataframeJSON = out.str()};
    pybx::Blotter::DFWUPC dfwupc{.df = df, .update_c = -2};
    for (auto it = rops.begin(); it != rops.end(); ) {
      try {
	(*it).oneway = true;
	(*it).show(dfwupc);
	it++;
      } catch (pybx::BadROP& e) {
	cout << "caught pybx::BadROP: " << e.what() << endl;
	cout << "removing subscriber at pos " << (it - rops.begin()) << endl;
	it = rops.erase(it);
	cout << "new rops size: " << rops.size() << endl;
      }
    }
  }
};

template <> inline StructDescriptor get_struct_descriptor<DFTestI::Rec>()
{
  static const StructDescriptor sd{
    make_member_descriptor("city", &DFTestI::Rec::city),
      make_member_descriptor("state", &DFTestI::Rec::state),
      make_member_descriptor("chance_of_rain", &DFTestI::Rec::chance_of_rain)
      };
  return sd;
} 

ADD_ACTION("test_subscriber[]", [](const Fuargs::args&) {
    pybx::Communicator comm;
    auto testdf_rop = comm.get_rop<pybx::Blotter::DFTest>("ws://localhost:8080/", "test_df");
    auto subscriber_o = make_shared<ObserverI>();
    auto subscriber_rop = comm.add_object<pybx::Blotter::Observer>(subscriber_o, "test_df");
    testdf_rop.subscribe(subscriber_rop);
    comm.run();
    return true;
  });

ADD_ACTION("test[]", [](const Fuargs::args&) {
    pybx::Communicator comm;
    auto testdf_rop = comm.get_rop<pybx::Blotter::DFTest>("ws://localhost:8080/", "test_df");
    auto df = testdf_rop.get_df();
    cout << "df: " << df.df.dataframeJSON << endl;
    return true;
  });

ADD_ACTION("run[]", [](const Fuargs::args&) {
    int port = 8080;
    pybx::Communicator comm;
    comm.set_listen_port(port);

    auto dftest_o = make_shared<DFTestI>();
    thread t(bind(&DFTestI::update_thread, dftest_o));

    comm.add_object<pybx::Blotter::DFTest>(dftest_o, "test_df");    
    comm.run();

    return true;
  });

int main(int argc, char** argv)
{
  Fuargs::exec_actions(argc, argv, false, false);
}
